File: research/explorations/bench-realistic/hashmap.js
function bench(n) {
    let s = 0;
    for (let i = 0; i < n; i++) {
        const e = { k: i, v: i * 7, sum: 0 };
        const e2 = { ...e, sum: e.k + e.v };
        s += e2.sum;
    }
    return s;
}

const N = parseInt(process.argv[2]) || 1000;
for (let i = 0; i < 1000; i++) bench(N);
const iters = 10000;
const start = process.hrtime.bigint();
let r;
for (let i = 0; i < iters; i++) r = bench(N);
const elapsed = Number(process.hrtime.bigint() - start);
console.log(`result:     ${r}`);
console.log(`iterations: ${iters}`);
console.log(`total:      ${(elapsed / 1e6).toFixed(2)}ms`);
console.log(`per call:   ${Math.floor(elapsed / iters)}ns`);

File: research/explorations/bench-realistic/hashmap.rb
Entry = Struct.new(:k, :v, :sum)

def bench(n)
  s = 0
  n.times do |i|
    e = Entry.new(i, i * 7, 0)
    e2 = Entry.new(e.k, e.v, e.k + e.v)
    s += e2.sum
  end
  s
end

n = (ARGV[0] || 1000).to_i
100.times { bench(n) }
iters = 10000
start = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
r = nil
iters.times { r = bench(n) }
elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - start
puts "result:     #{r}"
puts "iterations: #{iters}"
puts "total:      #{'%.2f' % (elapsed / 1e6)}ms"
puts "per call:   #{elapsed / iters}ns"
